src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs

```diff
@@ -0,0 +1,252 @@
+        private readonly Dictionary<Slice, int> _knownFields;
+        private static readonly byte[] _trueLiteral = Encoding.UTF8.GetBytes("true"); 
+        private static readonly byte[] _falseLiteral = Encoding.UTF8.GetBytes("false"); 
+
+        public CoraxDocumentConverter(
+                Index index, 
+                bool indexImplicitNull = false, 
+                bool indexEmptyEntries = true, 
+                string keyFieldName = null, 
+                bool storeValue = false, 
+                string storeValueFieldName = Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName) :
+            base(index, storeValue)
+        {
+            _allocator = new ByteStringContext(SharedMultipleUseFlag.None);
+            _knownFields = GetKnownFields();
+        }
+
+        public Dictionary<Slice, int> GetKnownFields()
+        {
+            if (_knownFields != null)
+                return _knownFields;
+            var knownFields = new Dictionary<Slice, int>();
+            foreach (var field in _fields.Values)
+            {
+                if(field.Name == null) 
+                    continue;
+                
+                Slice.From(_allocator, field.Name, ByteStringType.Immutable, out var value);
+                knownFields.Add(value, field.Id);
+            }
+
+            return knownFields;
+        }
+        
+        public Span<byte> InsertDocumentFields(LazyStringValue key, LazyStringValue sourceDocumentId, object doc, JsonOperationContext indexContext, out LazyStringValue id)
+        {
+            var document = (Document)doc;
+            _allocator.Allocate(document.Data.Size + _fields.Count * 1024, out ByteString buffer);
+            var entryWriter = new IndexEntryWriter(buffer.ToSpan(), _knownFields);
+            id = document.LowerId;
+            bool shouldSkip;
+            
+            
+            //TODO maciej - please look at this.
+            // this is reference to list for EnumerableWritingScope due to making it persistence during indexing document. 
+            // We want avoid allocation for every enumerable in doc, but if think it should be persistence during whole indexing process.
+            // Where should we put this and when release it?
+            List<int> stringsLength = new List<int>(128);
+            var scope = new SingleEntryWriterScope(stringsLength);
+            foreach (var indexField in _fields.Values)
+            {
+                if (BlittableJsonTraverserHelper.TryRead(_blittableTraverser, document, indexField.OriginalName ?? indexField.Name, out var value) == true)
+                {
+                    InsertRegularField(indexField, value, indexContext, out shouldSkip, ref entryWriter, scope);
+                }
+            }
+
+            //todo maciej try to figure out how should I deal with shouldSkip
+            entryWriter.Finish(out var output);
+            _allocator.Release(ref buffer);
```

Comment:
Shouldn't be in try / finally? You can also wrap `_allocator.Allocate()` into `using`

Reply:
good point.